### PR TITLE
8310242: Clarify the name parameter to Class::forName

### PR DESCRIPTION
**Pull request by mlchung**
Hi all,

This pull request contains a backport of commit [7db2f087](https://github.com/openjdk/jdk/commit/7db2f08756b0aa1d79cdd2356ed42aa5ab8bc58b) from the [openjdk/jdk](https://git.openjdk.org/jdk) repository.

The commit being backported was authored by Mandy Chung on 26 Jun 2023 and was reviewed by Roger Riggs, Chen Liang, Alan Bateman and David Holmes.

Thanks!

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change requires CSR request [JDK-8310244](https://bugs.openjdk.org/browse/JDK-8310244) to be approved
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issues
 * [JDK-8310242](https://bugs.openjdk.org/browse/JDK-8310242): Clarify the name parameter to Class::forName (**Bug** - P4)
 * [JDK-8310922](https://bugs.openjdk.org/browse/JDK-8310922): java/lang/Class/forName/ForNameNames.java fails after being added by JDK-8310242 (**Bug** - P2)
 * [JDK-8310244](https://bugs.openjdk.org/browse/JDK-8310244): Clarify the name parameter to Class::forName (**CSR**)

### Reviewers
 * [David Holmes](https://openjdk.org/census#dholmes) (@dholmes-ora - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk21.git pull/65/head:pull/65` \
`$ git checkout pull/65`

Update a local copy of the PR: \
`$ git checkout pull/65` \
`$ git pull https://git.openjdk.org/jdk21.git pull/65/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 65`

View PR using the GUI difftool: \
`$ git pr show -t 65`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk21/pull/65.diff">https://git.openjdk.org/jdk21/pull/65.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk21/pull/65#issuecomment-1608203816)

### CODE REVIEW

File: test/jdk/java/lang/Class/forName/ForNameNames.java [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+import static org.junit.jupiter.api.Assertions.*;
+
+public class ForNameNames {
+    static class Inner {}
+    static Stream<Arguments> testCases() {
+        return Stream.of(
+                Arguments.of("java.lang.String", String.class),
+                Arguments.of("[Ljava.lang.String;", String[].class),
+                Arguments.of("ForNameNames$Inner", Inner.class),
+                Arguments.of("[LForNameNames$Inner;", Inner.class),
```
Comment (liach) [inline test/jdk/java/lang/Class/forName/ForNameNames.java:46]: JDK-8310922 needs to be backported too.

Comment (mlchung) [reply test/jdk/java/lang/Class/forName/ForNameNames.java:46]: Yup... working on it.

**Review by dholmes-ora — APPROVED** [review]
Looks good. Thanks.

**Pull request closed without merging**